### PR TITLE
[feature] add volume files from secret

### PR DESCRIPTION
**Pull request by barloc**
Add volume mount from secret to the main container from dictionary
`{"secret_name": {"secret_key": "path_to_the_mount_point"}`

### CODE REVIEW

**Comment by LamaAni** [comment]
Sorry for the late response. Was not around.

File: airflow_kubernetes_job_operator/kubernetes_job_operator.py
Comment (LamaAni) [inline airflow_kubernetes_job_operator/kubernetes_job_operator.py]: Can you please update the help with more info here.

File: airflow_kubernetes_job_operator/kubernetes_job_operator.py
Comment (LamaAni) [inline airflow_kubernetes_job_operator/kubernetes_job_operator.py:287]: Problem here that this will be updated just to the main container. Hum.. this is ok but has caviats. Please update the help to reflect this. If should clearly state that this updates the main container only.

File: airflow_kubernetes_job_operator/kubernetes_job_operator.py
Comment (LamaAni) [reply airflow_kubernetes_job_operator/kubernetes_job_operator.py]:
```suggestion
            volume_files_from_secret (dict, optional): A dictionary of dictionaries of { 'secret_name': {'secret_key': 'load filepath'}}. Loads secrets from a Kubernetes secrets collection, to specific paths within the main pod (Volume from secret). 
```

File: airflow_kubernetes_job_operator/kubernetes_job_operator.py
Comment (LamaAni) [inline airflow_kubernetes_job_operator/kubernetes_job_operator.py]: Consider changing name: mount_files_from_secret? or secret_mounted_files? or something alike. Not sure this is clear enough.

File: airflow_kubernetes_job_operator/kubernetes_job_operator.py
Comment (LamaAni) [inline airflow_kubernetes_job_operator/kubernetes_job_operator.py]: Maybe this should be a list from a specific type? e.g. custom class that maps the secret config? 
i.e. 
```python
class FilesFromSecret(dict):
  def __init__(self, secret_name:str, **kwargs):
   super().__init__(**kwargs)
     self.secret_name=secret_name
```

And then the input argument is 
```python
volume_files_from_secret: Union[FilesFromSecret, List[FilesFromSecret]] = None
```
Thus allowing someone to send either a single secret or a list of them (initialized as list in __init__ if a single secret load).

File: airflow_kubernetes_job_operator/kubernetes_job_operator.py
Comment (LamaAni) [inline airflow_kubernetes_job_operator/kubernetes_job_operator.py]: This should be applied before the main container spec is applied, so future commands/derived commands can search for this. Please also add a comment.

**Review by LamaAni — CHANGES_REQUESTED** [review]
Nice, this definitely adds functionality. Please see my comments.

File: airflow_kubernetes_job_operator/kubernetes_job_operator.py
Comment (barloc) [reply airflow_kubernetes_job_operator/kubernetes_job_operator.py]: I thought about it but there are custom classes in the dag when I want to use it feature. And I think it is not good idea.
Add more docs and samples is simpler.

Comment (LamaAni) [reply airflow_kubernetes_job_operator/kubernetes_job_operator.py]: OK In that case, please add proper documentation for it so its clear what it does.

**Comment by barloc** [comment]
add some info and rename param

File: airflow_kubernetes_job_operator/kubernetes_job_operator.py
Comment (LamaAni) [inline airflow_kubernetes_job_operator/kubernetes_job_operator.py]: Change name to `_to_kubernetes_pod_secrets_file_mount_volume_list`?

File: airflow_kubernetes_job_operator/kubernetes_job_operator.py
Comment (LamaAni) [inline airflow_kubernetes_job_operator/kubernetes_job_operator.py]: See below the secret name in the spec volume generation. This name must be generated with some prefix as to no override other secrets coming from the template.

File: airflow_kubernetes_job_operator/kubernetes_job_operator.py
Comment (LamaAni) [inline airflow_kubernetes_job_operator/kubernetes_job_operator.py]: Please prepend the volume name with some autogenerated value, so the file mounted volumes will not clash with custom templates. e.g., if the template is:
```yaml
spec:
   volumes:
       my_secret: {}
```
and you add another, `my_secret` via the file load this would fail. (Duplicate value). So in this case we can add:
```yaml
                    *[{"name": "__auto_mounted_secret_files_"+k, "secret": {"secretName": k}} for k, _ in self.mount_files_from_secret.items()],
```

You can also make a function `_compose_auto_generated_file_mount_volume_secret_name(self, secret_name)` which will make life easier if someone really wants to mess with it.

**Review by LamaAni — CHANGES_REQUESTED** [review]
Hi sorry, some more comments I just noticed. Appologies.

File: airflow_kubernetes_job_operator/kubernetes_job_operator.py
Comment (barloc) [reply airflow_kubernetes_job_operator/kubernetes_job_operator.py]: Oh, good idea

File: airflow_kubernetes_job_operator/kubernetes_job_operator.py
Comment (LamaAni) [reply airflow_kubernetes_job_operator/kubernetes_job_operator.py]: It also needs to be unique (the secret name must be loaded only once :))

File: airflow_kubernetes_job_operator/kubernetes_job_operator.py
Comment (barloc) [reply airflow_kubernetes_job_operator/kubernetes_job_operator.py]: > This name must be generated with some prefix as to no override other secrets coming from the template.

This name we get from the input dictionary and it is [unique](https://kubernetes.io/docs/concepts/overview/working-with-objects/names/) in that namespace. So if somebody add this name to the template, secret will be rewritten.
I propose to add some info about it to the readme.
How about it?

Comment (LamaAni) [reply airflow_kubernetes_job_operator/kubernetes_job_operator.py]: Hum... but the volume is not the secret. That is how the final yaml would look when mounting the volumes (line 303 below)
```yaml
volumes:
    - name: [dependent on the deployment and can be multiple pointing to the same secret]
      secret:
        secretName: [always the sname]
```

e.g. the volume name cannot be the secret name. I'd suggest something like:
```yaml
volumes:
    - name: __kube_op_file_mount_secret_name
      secret:
        secretName: secret_name
```

Therefore the secret in this case must be mounted with a proper unique volume name, e.g.:
```yaml
volumeMounts:
        - name: __kube_op_file_mount_secret_name
          readOnly: true
          mountPath: the_mount_path
          subPath: secret_key
```

(Also best to load secrets with read_only by default, this extended properties is the reason I was thinking a class, but a Operator argument would work just as well)

Comment (LamaAni) [reply airflow_kubernetes_job_operator/kubernetes_job_operator.py]: From here: https://kubernetes.io/docs/concepts/configuration/secret/

And appologies for the late response was on vacation

Comment (barloc) [reply airflow_kubernetes_job_operator/kubernetes_job_operator.py]: `volumes:
    - name: [dependent on the deployment and can be multiple pointing to the same secret]
      secret:
        secretName: [always the sname]`

Yep, it is true. 
But secretName is unique always, so I don't understand why use not secret name for volume naming :)

An usage of the `__kube_op_` or other prefix - not helps, because somebody do this naming sooner or later :)

Comment (LamaAni) [reply airflow_kubernetes_job_operator/kubernetes_job_operator.py]: I seem to be missing something. From the link, the pod volume mounts are as follows:
```yaml
apiVersion: v1
kind: Secret
metadata:
  name: dotfile-secret
data:
  .secret-file: dmFsdWUtMg0KDQo=
---
apiVersion: v1
kind: Pod
metadata:
  name: secret-dotfiles-pod
spec:
  volumes:
    - name: secret-volume
      secret:
        secretName: dotfile-secret
  containers:
    - name: dotfile-test-container
      image: registry.k8s.io/busybox
      command:
        - ls
        - "-l"
        - "/etc/secret-volume"
      volumeMounts:
        - name: secret-volume
          readOnly: true
          mountPath: "/etc/secret-volume"
```

How would your yaml look like?

Also it seems I need to add a print yaml mode to the operator to allow to print the final yaml result so we'd have less confusion. Hum... :)

**Comment by LamaAni** [comment]
Hi, sorry for the late reply. Was traveling. Any progress on this?

**Comment by LamaAni** [comment]
Hi, sorry for the late reply. Was traveling. Any progress on this?

**Comment by barloc** [comment]
> Hi, sorry for the late reply. Was traveling. Any progress on this?

Hello.
I fix all what you say :)

**Review by LamaAni — APPROVED** [review]

**Comment by LamaAni** [comment]
Approved sorry for the delay. I'm traveling

**Comment by barloc** [comment]
Good news :) Waiting for merge )

**Comment by LamaAni** [comment]
You can merge the PR's sorry was traveling again. Back in the states now and things should go faster. Once the other PR is merged I will release a version.

**Pull request merged**
